File: src/core/validators/MergePartOfReference/MergePartOfReferenceExistsOnlyInCoreNamespace.js
// @flow
import R from 'ramda';
import { errorRuleBase } from '../ValidationRuleBase';
import { includeRuleBase } from '../ValidationRuleRepository';
import { MetaEdGrammar } from '../../../grammar/gen/MetaEdGrammar';
import type SymbolTable from '../SymbolTable';
import { entityIdentifier, entityName } from '../RuleInformation';
import { namespaceAncestorContext, topLevelEntityAncestorContext, propertyAncestorContext,
  isExtensionNamespace, namespaceNameFor } from '../ValidationHelper';

// eslint-disable-next-line no-unused-vars
export function valid(ruleContext: any, symbolTable: SymbolTable): boolean {
  return R.compose(R.not, isExtensionNamespace, namespaceAncestorContext)(ruleContext);
}

// eslint-disable-next-line no-unused-vars
function failureMessage(ruleContext: any, symbolTable: SymbolTable): string {
  const namespaceName = namespaceNameFor(namespaceAncestorContext(ruleContext));
  const parentEntity = topLevelEntityAncestorContext(ruleContext);
  const propertyName = propertyAncestorContext(ruleContext).propertyName().ID().getText();
  return `'merge' is invalid for property ${propertyName} on ${entityIdentifier(parentEntity)} '${entityName(parentEntity)}' in extension namespace ${namespaceName}.  'merge' is only valid for properties on types in a core namespace.`;
}

const validationRule = errorRuleBase(valid, failureMessage);
// eslint-disable-next-line import/prefer-default-export
export const includeRule = includeRuleBase(MetaEdGrammar.RULE_mergePartOfReference, validationRule);

File: src/core/validators/MergePartOfReference/MergePartOfReferenceExistsOnlyInCoreNamespace.cs
using MetaEd.Grammar.Antlr;
using MetaEd.Grammar.Antlr.Extensions;

namespace MetaEd.Core.Validator.MergePartOfReference
{
    public class MergePartOfReferenceExistsOnlyInCoreNamespace : ValidationRuleBase<MetaEdGrammar.MergePartOfReferenceContext>
    {
        public override bool IsValid(MetaEdGrammar.MergePartOfReferenceContext context)
        {
            var namespaceInfo = context.GetAncestorContext<INamespaceInfo>();
            return !namespaceInfo.IsExtension;
        }

        public override string GetFailureMessage(MetaEdGrammar.MergePartOfReferenceContext context)
        {
            var namespaceInfo = context.GetAncestorContext<INamespaceInfo>();
            var topLevelEntity = context.GetAncestorContext<ITopLevelEntity>();
            var propertyWithComponents = context.GetAncestorContext<IPropertyWithComponents>();
            return
                string.Format(
                    "'merge' is invalid for property {0} on {1} '{2}' in extension namespace {3}.  'merge' is only valid for properties on types in a core namespace.",
                    propertyWithComponents.IdNode().GetText(),
                    topLevelEntity.EntityIdentifier(),
                    topLevelEntity.EntityName(),
                    namespaceInfo.NamespaceName);
        }
    }
}
